### PR TITLE
sys/bhp_msg: add IPC based implementation of Bottom Half Processor

### PR DESCRIPTION
**Pull request by jia200x**
### Contribution description
This PR adds an IPC version of the `bhp` module. It allows device independent IRQ offloading using message queues.
A `bhp_msg_t` descriptor holds the PID of a thread that performs Bottom Half processing on reception of a message with a dedicated type. Similar to https://github.com/RIOT-OS/RIOT/pull/18435, this mechanism can be adapted to existing `init` functions of drivers to hook in IRQ offloading.

### Testing procedure
`tests/unittests` should pass.

### Issues/PRs references
https://github.com/RIOT-OS/RIOT/pull/18435
Required for adding LWIP support for radios that require IRQ offloading

### CODE REVIEW

**Comment by kaspar030** [comment]
This is a bad idea. Messages sent from ISR *silently vanish* if the receiver is not waiting (in `msg_receive()`), or if there's a message queue and that is full. Both will happen once there's load, so it's difficult to track down.

The effect would be easy to show with a unit test that calls `bhp_msg_isr_cb()` twice from an actual ISR. Without a queue, the second message will be lost. Or with a queue of size N, call bhp_msg_isr_cb` N+1 times.

I'm beginning to think we should deprecate `msg_send()` from ISRs and only allow `msg_try_send()`, which `msg_send()` will use anyways when called from ISR, eating the result code.

**Comment by jia200x** [comment]
As discussed in many old PRs, I was never a fan of ISR offloading using IPC and  I 100% agree with what you describe. I plan to migrate all the remaining IPC based IRQ Offloading modules to Events or Thread Flags at some point.
I only introduce this module to give support for the _existing_ modules (concretely LWIP), since they already use IPC.
IMO it shouldn't make the situation worse but at least it should speed up the Radio HAL migration.

**Comment by benpicco** [comment]
OK I see, this is needed because LWIP uses msg already for netdev events and this will integrate with it.

File: sys/bhp/Kconfig [hunk 1]
```diff
@@ -22,4 +22,9 @@ config MODULE_BHP_EVENT
     depends on MODULE_EVENT
     depends on MODULE_BHP
 
+config MODULE_BHP_MSG
+    bool "Enable IPC based Bottom Half Processor implementation"
```
Comment (benpicco) [inline sys/bhp/Kconfig:26]:
```suggestion
    bool "Enable message based Bottom Half Processor implementation"
```

**Review by benpicco — COMMENTED** [review]
Code looks good, but there should be some explanation as to when `bhp_event` and when `bhp_msg` should be used.

CI has some comments too.

**Comment by jia200x** [comment]
done!

**Review by benpicco — COMMENTED** [review]
Static tests are still failing

**Comment by benpicco** [comment]
Please squash!

**Comment by jia200x** [comment]
squashed!

File: tests/unittests/tests-bhp_msg/tests-bhp_msg.c [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+}
+
+static void *_event_loop(void *arg)
+{
+    (void) arg;
+    while (1) {
+        msg_t msg;
+        msg_receive(&msg);
+        if (msg.type == BHP_MSG_BH_REQUEST) {
+            bhp_irq_handler(msg.content.ptr);
```
Comment (benpicco) [inline tests/unittests/tests-bhp_msg/tests-bhp_msg.c:58]: Maybe add a helper function

```C
static inline void bhp_msg_handle(msg_t *msg)
{
    assert(msg->type == BHP_MSG_BH_REQUEST);
    bhp_irq_handler(msg->content.ptr);
}
```

so it's more obvious what to do with the msg

Comment (benpicco) [reply tests/unittests/tests-bhp_msg/tests-bhp_msg.c:58]: Hm I suppose that's a no?

Comment (jia200x) [reply tests/unittests/tests-bhp_msg/tests-bhp_msg.c]: done. I amended it directly.

**Review by benpicco — APPROVED** [review]

**Pull request merged**